grpc/src/main/java/com/linecorp/armeria/common/grpc/GsonGrpcJsonMarshallerBuilder.java

```diff
@@ -30,42 +34,117 @@
  * to and from JSON.
  */
 public final class GsonGrpcJsonMarshallerBuilder {
+    private static final Logger logger = LoggerFactory.getLogger(GsonGrpcJsonMarshallerBuilder.class);
+
+    private static boolean loggedJsonParserCustomizerWarning;
+    private static boolean loggedJsonPrinterCustomizerWarning;
 
     @Nullable
-    private Consumer<JsonFormat.Parser> jsonParserCustomizer;
+    private Function<JsonFormat.Parser, JsonFormat.Parser> jsonParserCustomizer;
 
     @Nullable
-    private Consumer<JsonFormat.Printer> jsonPrinterCustomizer;
+    private Function<JsonFormat.Printer, JsonFormat.Printer> jsonPrinterCustomizer;
 
     GsonGrpcJsonMarshallerBuilder() {}
 
+    /**
+     * Adds a {@link Function} that returns customized the {@link JsonFormat.Parser}
+     * used when deserializing a JSON payload into a {@link Message}.
+     */
+    public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
+            Function<? super JsonFormat.Parser, JsonFormat.Parser> jsonParserCustomizer) {
+        requireNonNull(jsonParserCustomizer, "jsonParserCustomizer");
+        if (this.jsonParserCustomizer == null) {
+            @SuppressWarnings("unchecked")
+            final Function<JsonFormat.Parser, JsonFormat.Parser> cast =
+                    (Function<JsonFormat.Parser, JsonFormat.Parser>) jsonParserCustomizer;
+            this.jsonParserCustomizer = cast;
+        } else {
+            this.jsonParserCustomizer = this.jsonParserCustomizer.andThen(jsonParserCustomizer);
+        }
+        return this;
+    }
+
     /**
      * Adds a {@link Consumer} that can customize the {@link JsonFormat.Parser}
      * used when deserializing a JSON payload into a {@link Message}.
+     *
+     * @deprecated {@link JsonFormat.Parser} is immutable so all changes applied in the {@link Consumer}
+     *     will be lost. Please use the {@link #jsonParserCustomizer(Function) jsonParserCustomizer}
+     *     which accepts {@link Function} parameter instead.
      */
+    @Deprecated
     public GsonGrpcJsonMarshallerBuilder jsonParserCustomizer(
             Consumer<? super JsonFormat.Parser> jsonParserCustomizer) {
+        if (!loggedJsonParserCustomizerWarning) {
+            logger.warn("{}.jsonParserCustomizer(Consumer) does not work as expected, " +
+                            "use jsonParserCustomizer(Function).",
+                    getClass().getSimpleName());
```

Comment:
Let me fix the indentation:
```suggestion
            logger.warn("{}.jsonParserCustomizer(Consumer) does not work as expected, " +
                        "use jsonParserCustomizer(Function).",
                        getClass().getSimpleName());
```